File: apps/api/src/test/java/br/unifor/costify/application/usecase/RegisterIngredientUseCaseTest.java
package br.unifor.costify.application.usecase;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import br.unifor.costify.application.contracts.IngredientRepository;
import br.unifor.costify.application.dto.command.RegisterIngredientCommand;
import br.unifor.costify.application.dto.entity.IngredientDto;
import br.unifor.costify.application.errors.IngredientAlreadyExistsException;
import br.unifor.costify.application.factory.IngredientFactory;
import br.unifor.costify.application.validation.ValidationService;
import br.unifor.costify.domain.contracts.IdGenerator;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.Unit;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

class RegisterIngredientUseCaseTest {

  @Mock private IngredientRepository ingredientRepository;

  @Mock private IdGenerator idGenerator;

  @Mock private ValidationService validationService;

  private RegisterIngredientUseCase useCase;
  private IngredientFactory ingredientFactory;

  @BeforeEach
  void setup() {
    MockitoAnnotations.openMocks(this);
    ingredientFactory = new IngredientFactory(idGenerator);
    useCase = new RegisterIngredientUseCase(ingredientRepository, ingredientFactory, validationService);
  }

  @Test
  void execute_withValidCommand_shouldSaveAndReturnIngredientDto() {
    RegisterIngredientCommand command = new RegisterIngredientCommand("Flour", 1.0, 5.0, Unit.KG);

    when(ingredientRepository.existsByName("Flour")).thenReturn(false);
    when(idGenerator.generate()).thenReturn("test-id-123");

    Ingredient savedIngredient = new Ingredient(idGenerator, "Flour", 1.0, Money.of(5.0), Unit.KG);
    when(ingredientRepository.save(any(Ingredient.class))).thenReturn(savedIngredient);

    IngredientDto result = useCase.execute(command);

    assert result.name().equals("Flour");
    assert result.packageQuantity() == 1.0;
    assert result.packagePrice() == 5.0;
    assert result.packageUnit() == Unit.KG;

    verify(ingredientRepository).existsByName("Flour");
    verify(ingredientRepository).save(any(Ingredient.class));
  }

  @Test
  void execute_withExistingIngredientName_shouldThrowException() {
    RegisterIngredientCommand command = new RegisterIngredientCommand("Flour", 1.0, 5.0, Unit.KG);

    when(ingredientRepository.existsByName("Flour")).thenReturn(true);

    try {
      useCase.execute(command);
      assert false;
    } catch (IngredientAlreadyExistsException e) {
      assert e.getMessage().contains("Flour");
      assert e.getMessage().contains("already exists");
    }

    verify(ingredientRepository).existsByName("Flour");
    verify(ingredientRepository, never()).save(any(Ingredient.class));
  }

  @Test
  void execute_withValidCommand_shouldCallRepositoryMethods() {
    RegisterIngredientCommand command = new RegisterIngredientCommand("Sugar", 2.0, 8.0, Unit.KG);

    when(ingredientRepository.existsByName("Sugar")).thenReturn(false);
    when(idGenerator.generate()).thenReturn("sugar-id-456");

    Ingredient savedIngredient = new Ingredient(idGenerator, "Sugar", 2.0, Money.of(8.0), Unit.KG);
    when(ingredientRepository.save(any(Ingredient.class))).thenReturn(savedIngredient);

    useCase.execute(command);

    verify(ingredientRepository, times(1)).existsByName("Sugar");
    verify(ingredientRepository, times(1)).save(any(Ingredient.class));
  }
}

File: test/application/usecase/RegisterIngredientUseCase.test.ts
import { describe, expect, it, beforeEach, vi } from 'vitest'
import { RegisterIngredientUseCase } from '@application/usecase/RegisterIngredientUseCase.js'
import type { IngredientRepository } from '@application/contracts/IngredientRepository.js'
import type { IdGenerator } from '@domain/contracts/IdGenerator.js'
import { Ingredient } from '@domain/entities/Ingredient.js'
import { Unit } from '@domain/valueobjects/Unit.js'
import { IngredientAlreadyExistsException } from '@application/errors/IngredientAlreadyExistsException.js'

describe('RegisterIngredientUseCase', () => {
  let useCase: RegisterIngredientUseCase
  let mockRepository: IngredientRepository
  let mockIdGenerator: IdGenerator

  beforeEach(() => {
    mockRepository = {
      save: vi.fn(),
      findById: vi.fn(),
      findByName: vi.fn(),
      findAll: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
      existsById: vi.fn(),
      existsByName: vi.fn(),
    }

    mockIdGenerator = {
      generate: vi.fn(() => '550e8400-e29b-41d4-a716-446655440000')
    }

    useCase = new RegisterIngredientUseCase(mockRepository, mockIdGenerator)
  })

  it('should register new ingredient successfully', async () => {
    const command = {
      name: 'Flour',
      pricePerUnit: '2.50',
      unit: Unit.KILOGRAM
    }

    vi.mocked(mockRepository.findByName).mockResolvedValue(null)

    const result = await useCase.execute(command)

    expect(result).toEqual({
      id: '550e8400-e29b-41d4-a716-446655440000',
      name: 'Flour',
      pricePerUnit: '2.50',
      unit: Unit.KILOGRAM
    })
    expect(mockRepository.save).toHaveBeenCalledOnce()
  })

  it('should throw error when ingredient already exists', async () => {
    const command = {
      name: 'Flour',
      pricePerUnit: '2.50',
      unit: Unit.KILOGRAM
    }

    const existingIngredient = new Ingredient(
      { getValue: () => 'existing-id' } as any,
      'Flour',
      { toFixed: () => '2.50' } as any,
      Unit.KILOGRAM
    )

    vi.mocked(mockRepository.findByName).mockResolvedValue(existingIngredient)

    await expect(useCase.execute(command))
      .rejects.toThrow(IngredientAlreadyExistsException)

    expect(mockRepository.save).not.toHaveBeenCalled()
  })
})